### ISSUE TITLE
Rule proposal: no-unsafe-assertion (no explicit casts from any/unknown)

### ISSUE THREAD

**Issue by inga-lovinde** [typescript-eslint/typescript-eslint#8723]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

This rule would prevent dangerous type assertion / casts from unvalidated data.

It is unsafe to simply assert that just any arbitrary JSON will conform to the expected type, this can cause issues in runtime (e.g. `const data = JSON.parse("{}") as { a: string }; console.log(data.a.length);`); it is safer to validate them first.

There are many validation libraries frameworks that create schemas from TS types or TS types from schemas, validate input data according to that schema, and return the validated data already appropriately typed (e.g. zod, tsoa, typebox).

In codebases that use such libraries, it can be very useful to identify places where unsafe casts / type assertions are done, or for linting to help avoid accidentally making such unsafe casts.

### Fail Cases

```TypeScript
const data = response.json() as MyDataType;
// this call will throw e.g. if response.json() returns {}
console.log(data.myField.myOtherField);
```

```TypeScript
app.post('/', (req, res, next) => {
    const data = req.body as MyBodyModel;
    // this call will throw e.g. if body is {}
    console.log(data.myField.myOtherField);
});
```

```TypeScript
const trimmed = (anyData as string).trim();
```

### Pass Cases

```TypeScript
// zod throws a detailed validation error when data doesn't match the schema
const data = myDataTypeZodSchema.parse(response.json());
// after that, the data can be used safely, this call will never cause errors
console.log(data.myField.myOtherField);
```

```TypeScript
// when data doesn't match the schema, tsoa will return HTTP Bad Request by default, and will not call the method
@Post('/')
public async Post(@Body() data: MyBodyModel) {
    // so inside of this method, the data can be used safely, this call will never cause errors
    console.log(data.myField.myOtherField);
}
```

```TypeScript
if (typeof anyData !== "string") {
    // handle incorrect data; throw or return early
}

// anyData is narrowed down to string by the condition above
const trimmed = anyData.trim();
```

**Comment by JoshuaKGoldberg** [comment]
Heh, I've definitely been in projects where this kind of rule would have been wonderful... unfortunately, using it would be _very_ restrictive. Even in projects that use a helper library like Zod, it's common that you might want to `JSON.parse`/etc. in the occasional edge case.

My hunch is that this would be best implemented as a shared community plugin so folks can iterate quickly on questions like:

* What, if any, exclusions would this rule allow for?
* Should it separate `as any` vs. `as unknown`?
    * Speaking of which: is this a superset of #7173?
* Practically, how many projects could actually adopt this without very many inline disable comments?

If we were to build this into typescript-eslint I don't think we'd be able to enable this in `strict` even to start.

...but maybe I'm just being pessimistic. `evaluating community engagement` it is!

**Comment by bradzacher** [comment]
I had the same thoughts myself.
The idea itself is grand but the reality is that the only way out of `any` is via an assertion.

`unknown` at least can be refined out of but not `any`. Which means that if you do get an `any` somehow (eg via `JSON.parse`) then you're SOL and will have to have an assertion and thus have to have a disable comment.

I could see a version of the rule that only checks `unknown` by default and has a (default off) option to check `any`. That is something that could be recommended.

**Comment by Josh-Cena** [comment]
> the only way out of `any` is via an assertion.

Not really; it can also be a type guard (such as validation library). I assume `any as unknown` would also be considered safe.

**Comment by bradzacher** [comment]
A type guard is a function and thus would be flagged by `no-unsafe-argument` unless the argument is explicitly typed as `any` or `unknown`.

**Comment by Josh-Cena** [comment]
> A type guard is a function and thus would be flagged by `no-unsafe-argument` unless the argument is explicitly typed as `any` or `unknown`.

Of course a type guard would be typed as accepting `unknown`—anything else is blatantly unsafe! Validation libraries' `parse` methods all work like that.

> That's still an assertion so my statement still stands 😅

This issue wants to forbid `any as T` where `T` is not a top type because that would effectively be a downcast, but `any as unknown` is not a downcast and is not "unsafe" by any definition, so it would be allowed like how other "unsafe" rules allow `any -> unknown` comparisons.

**Comment by inga-lovinde** [comment]
Yes, #7173 looks somewhat related, I didn't notice it when I was looking for similar issues. But it's somewhat different, too; it requires more type data (TS assignability API is mentioned in the comments), and while it would disallow many things this suggestion would allow, it would still allow assignments from any to a specific type.

I don't think inline disable comments are inherently bad, for me they're more like of a way to communicate that yes, we are aware that dangerous things are going here, we're doing it on purpose and carefully and knowing that we have to pay extra attention here.
There is a compiler setting blocking implicit anys; there is a typescript-eslint rule blocking explicit anys (unless overridden); the only missing component is how to deal with explicit anys that came from the other code, so even though they might be explicit where they're implemented, they become implicit where they're used.

And also of course it should allow casts to unknown; `JSON.parse(...) as unknown` is perfectly safe after all, as opposed to `JSON.parse(...) as MyDataType`.

I don't think this should be a part of a strict ruleset, it can be a separate rule on its own (as many rules in typescript-eslint are, AFAIK?)
And I cannot think of any configuration options at the moment, it's just one rule, disallow casting something that is of type `any` to anything specific (i.e. anything besides `any` or `unknown`).

**Comment by JoshuaKGoldberg** [comment]
> the only missing component is how to deal with explicit anys that came from the other code

Just confirming: do you think this rule would still be necessary for folks using the `no-unsafe-*` family of rules around `any`s? E.g. [`@typescript-eslint/no-unsafe-argument](https://typescript-eslint.io/rules/no-unsafe-argument)

**Comment by inga-lovinde** [comment]
@JoshuaKGoldberg , I'm using these rules, they prevent using values of `any` type without explicitly casting them, but they don't prevent these explicit casts (which can often be unsafe, as with `JSON.parse`d arbitrary input string).

**Comment by bradzacher** [comment]
5 months later and this request has received 3 community reactions.
As per our [workflow guidelines](https://typescript-eslint.io/maintenance/issues#community-engagement-evaluation) - we are taking this as signal that this check is not important to the broader community and are thus rejecting the proposal. Thanks for filing!
